### PR TITLE
Integrate Healthiness into Summarizer

### PR DESCRIPTION
**Pull request by itsazhuhere**
This PR does the following:
- Integrates healthiness calculations into Summarizer.
- Adds trend analysis by calculating the previous interval's healthiness
- Adds a field to config proto to opt in to doing healthiness calculations for a tab
- Adds unit tests for new code, as well as a unit test for an uncaught edge case in existing code.

### CODE REVIEW

**Comment by itsazhuhere** [comment]
Holding until I get the internal config.proto changed to match.

File: pb/config/config.proto [hunk 1]
```diff
@@ -578,6 +578,8 @@ message DashboardTab {
   // and will opt you into the beta AutoBug.
   AutoBugOptions beta_autobug_options = 22;
 
+  // Options for the configuration of the flakiness analysis tool, on a per tab basis
+  HealthinessAnalysisOptions healthiness_analysis_options = 23;
```
Comment (michelle192837) [inline pb/config/config.proto:582]: nit: 'health' instead of 'healthiness', just for brevity?

File: pb/config/config.proto [hunk 2]
```diff
@@ -629,6 +631,20 @@ message Configuration {
   repeated DashboardGroup dashboard_groups = 3;
 }
 
+// A grouping of configuration options for the flakiness analysis tool.
+// Later configuration options could include the ability to choose different kinds of 
+// flakiness and choosing if and who to email a copy of the flakiness report.
+message HealthinessAnalysisOptions {
+  // Defaults to false; flakiness analysis is opt-in
+  bool should_analyze = 1;
```
Comment (michelle192837) [inline pb/config/config.proto:639]: I think we can infer if this is enabled by checking if this message is nil or not, so should_analyze isn't needed?

File: pb/summary/summary.proto [hunk 3]
```diff
@@ -76,6 +76,19 @@ message TestInfo {
 
   // The flakiness of the test, measured out of 100
   float flakiness = 8;
+
+  enum Trend{
+    NONE = 0;
```
Comment (michelle192837) [inline pb/summary/summary.proto:81]: nit: I think it's worth differentiating 'unknown' (we didn't set the trend value) and 'no change' (no change in trend occurred)?

File: pkg/summarizer/summary.go [hunk 4]
```diff
@@ -592,6 +608,38 @@ func latestGreen(grid *statepb.Grid, useFirstExtra bool) string {
...
+
+	healthiness := CalculateHealthiness(grid, &analyzer, oneInterval, now, tabName)
+	pastHealthiness := CalculateHealthiness(grid, &analyzer, twoIntervals, oneInterval, tabName)
+	CalculateTrend(healthiness, pastHealthiness)
+
+	return healthiness
+}
+
+func goBackDays(days int, currentTime time.Time) int {
+	// The primary usage of goBackDays is to get the time intervals for our flakiness report
```
Comment (michelle192837) [inline pkg/summarizer/summary.go:627]: nit: This comment can read more like the one for CalculateTrend ("gobackDays gets...")

File: pkg/summarizer/summary_test.go [hunk 5]
```diff
@@ -159,7 +165,7 @@ func TestUpdateDashboard(t *testing.T) {
 				}
 				return &fake.group, reader, nil
 			}
-			actual, err := updateDashboard(context.Background(), tc.dash, finder)
+			actual, err := updateDashboard(context.Background(), tc.dash, finder, realClock)
```
Comment (michelle192837) [inline pkg/summarizer/summary_test.go:168]: I don't think you're testing time here, just in getHealthinessForInterval, so I think you only need to pass in time/clock for that function and not for the higher-level ones?

File: pb/config/config.proto (hunk 2 above)
Comment (itsazhuhere) [reply pb/config/config.proto:639]: I was thinking that if I was going to add defaults to the default YAML then I would need a flag because the message would potentially be populated by the defaults. Although since that's not the case right now, should it be checking for nil now and later we add the should_analyze flag when we have defaulting functionality in?

Comment (michelle192837) [reply pb/config/config.proto:639]: Ah actually, good point, I was not thinking about how to disable this is set in defaults but someone wants to opt-out. 

In that case, just a small nit: `enable` or similar for the field name?

(An alternative is to say, you have to set a positive days_of_analysis or we assume you didn't want to enable this, so people can opt out with days_of_analysis = 0. But I don't think that's as clear or necessary, unlike some other options where that makes more sense).

File: pkg/summarizer/summary_test.go (hunk 5 above)
Comment (itsazhuhere) [reply pkg/summarizer/summary_test.go:168]: I agree that the only place time is tested/needs to be tested is in getHealthinessForInterval. I guess when I put it in the higher level functions I was thinking that we would need to test how getHealthinessForInterval operates within the flow of updateDashboard and updateTab. Is it sufficient to test that getHealthinessForInterval works in a vacuum and reasonably assume it works in the functions that use it? I did think it was messy introducing another parameter just to pass it all the way down to one function.

Comment (itsazhuhere) [reply pkg/summarizer/summary_test.go]: After an offline discussion with @krzyzacy, I'll make a future PR that refactors things so it's easier to test the shouldRunHealthiness block of updateTab. It would involve moving most of the functions in summarizer.go into a summarizer struct that holds members that we can easily inject dummy versions into. 

I took out the new parameter and made getHealthinessForInterval depend on a Time instead of a function that returns Time.

File: pkg/summarizer/flakiness.go [hunk 6]
```diff
@@ -46,6 +48,37 @@ func CalculateHealthiness(grid *state.Grid, analyzer flakinessAnalyzer, startTim
...
+	for _, test := range previousHealthiness.Tests {
+		previousFlakiness[test.DisplayName] = test.Flakiness
+	}
+
+	for i, test := range currentHealthiness.Tests {
+		var prevFlakiness float32
+		if value, ok := previousFlakiness[test.DisplayName]; ok {
+			prevFlakiness = value
+		}
+		currentHealthiness.Tests[i].ChangeFromLastInterval = getTrend(test.Flakiness, prevFlakiness)
```
Comment (krzyzacy) [inline pkg/summarizer/flakiness.go:68]: nit: this can be simplified to

```
                currentHealthiness.Tests[i].ChangeFromLastInterval = summarypb.None
                if value, ok := previousFlakiness[test.DisplayName]; ok {
			currentHealthiness.Tests[i].ChangeFromLastInterval = getTrend(test.Flakiness, prevFlakiness)
		}
		
```

**Comment by krzyzacy** [comment]
wanna squash the commits?

**Comment by itsazhuhere** [comment]
Yup, done!
